metricbeat/module/prometheus/query/data.go

```diff
@@ -0,0 +1,137 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package query
+
+import (
+	"encoding/json"
+	"strconv"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/metricbeat/mb"
+)
+
+// ArrayResponse is for "scalar", "string" type.
+type ArrayResponse struct {
+	Status string    `json:"status"`
+	Data   arrayData `json:"data"`
+}
+type arrayData struct {
+	ResultType string        `json:"resultType"`
+	Results    []interface{} `json:"result"`
+}
+
+// MapResponse is for "vector", "matrix" type from Prometheus Query API Request
+type MapResponse struct {
+	Status string  `json:"status"`
+	Data   mapData `json:"data"`
+}
+type mapData struct {
+	ResultType string      `json:"resultType"`
+	Results    []mapResult `json:"result"`
+}
+type mapResult struct {
+	Metric  map[string]string `json:"metric"`
+	Vector  []interface{}     `json:"value"`
+	Vectors [][]interface{}   `json:"values"`
+}
+
+func parseResponse(body []byte, pathConfig QueryConfig) ([]mb.Event, error) {
+	var events []mb.Event
+	converted, resultType, err := convertJSONToStruct(body)
+	if err != nil {
+		return events, err
+	}
+	switch resultType {
+	case "scalar", "string":
+		res := converted.(ArrayResponse)
+		events = append(events, mb.Event{
+			Timestamp: getTimestamp(res.Data.Results[0].(float64)),
+			MetricSetFields: common.MapStr{
+				"dataType":           resultType,
+				pathConfig.QueryName: convertToNumeric(res.Data.Results[1].(string)),
+			},
+		})
+	case "vector":
+		res := converted.(MapResponse)
+		for _, result := range res.Data.Results {
+			events = append(events, mb.Event{
+				Timestamp:    getTimestamp(result.Vector[0].(float64)),
+				ModuleFields: common.MapStr{"labels": result.Metric},
+				MetricSetFields: common.MapStr{
+					"dataType":           resultType,
+					pathConfig.QueryName: convertToNumeric(result.Vector[1].(string)),
+				},
+			})
+		}
+	case "matrix":
+		res := converted.(MapResponse)
+		for _, result := range res.Data.Results {
+			for _, vector := range result.Vectors {
+				events = append(events, mb.Event{
+					Timestamp:    getTimestamp(vector[0].(float64)),
+					ModuleFields: common.MapStr{"labels": result.Metric},
+					MetricSetFields: common.MapStr{
+						"dataType":           resultType,
+						pathConfig.QueryName: convertToNumeric(vector[1].(string)),
+					},
+				})
+			}
+		}
+	default:
+		return events, errors.New("Unknown resultType " + resultType)
+	}
+	return events, nil
+}
+
+func convertJSONToStruct(body []byte) (interface{}, string, error) {
```

Comment:
Avoid code that returns `interface{}`. If the logic is complex split it in more functions. The `string` returned parameter is hidden from the title of the function. ie: According to title, why there's a string when I get a Struct? I'm actually getting an `interface{}` in fact